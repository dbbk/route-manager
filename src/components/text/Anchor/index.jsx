// @flow

import * as React from 'react';
import styled, { css } from 'styled-components';
import { switchProp, prop, ifProp } from 'styled-tools';
import { darken } from 'polished';

type Props = {
  anchorRef?: () => mixed,
  className?: string,
  children?: React.Node,
  element: any,
  email?: string,
  external?: boolean,
  hash?: string,
  href?: string,
  itemProp?: string,
  onTabFocus?: (e: SyntheticKeyboardEvent<*>) => mixed,
  onClick?: (e: SyntheticMouseEvent<*>) => mixed,
  target?: string,
  tel?: string,
  to?: string,
  variant?: 'standard' | 'inherit' | 'hyperlink' | 'navigation',
};

const TAB_KEY_CODE = 9;

function AnchorComponent(props: Props) {
  // `element` is renamed here in order to invoke it as a JSX component.
  const {
    anchorRef,
    className,
    children,
    element: ComposedElement,
    email,
    external,
    hash,
    itemProp,
    onClick,
    onTabFocus,
    target,
    tel,
    to,
  } = props;

  // The browser and React don't have an event for an element becoming focused
  // ONLY as a result of keyboard tabbing, so we must listen for the tab
  // key-up-press event on the element to call the `onTabFocus` callback.
  const onKeyUp = e =>
    e.keyCode === TAB_KEY_CODE && typeof onTabFocus === 'function'
      ? onTabFocus(e)
      : true;

  if (hash) {
    return (
      <a
        ref={anchorRef}
        href={`#${hash}`}
        className={className}
        onClick={onClick}
        onKeyUp={onKeyUp}
      >
        {children}
      </a>
    );
  } else if (tel) {
    return (
      <a href={`tel:${tel}`} className={className} onClick={onClick}>
        {children}
      </a>
    );
  } else if (email) {
    return (
      <a href={`mailto:${email}`} className={className} onClick={onClick}>
        {children}
      </a>
    );
  } else if (external) {
    return (
      <a
        href={to}
        className={className}
        target={target}
        rel="noopener noreferrer"
        onClick={onClick}
        itemProp={itemProp}
      >
        {children}
      </a>
    );
  }

  return <ComposedElement {...props}>{children}</ComposedElement>;
}

/*
TODO: The standard variant is the pre-rebranding style, and the hyperlink variant
is the post-rebranding style. Once rebranding has been fully rolled out, the
hyperlink styles should replace standard.
*/
const Anchor = styled(AnchorComponent)`
  cursor: pointer;
  text-decoration: none;
  ${switchProp('variant', {
    standard: css`
      color: ${prop('theme.colors.brand.primary')};`,
    navigation: css`
      display: inline-block;
      white-space: nowrap;
      font-weight: ${prop('theme.fontWeight.bold')};
      padding: 25px 0 20px;
      border-bottom: 5px solid transparent;
      margin-right: 40px;

      ${ifProp(
        { active: true },
        css`
          color: ${prop('theme.colors.greyscale.black')};
          border-bottom-color: ${prop('theme.colors.brand.primary')};
        `,
        css`
          color: ${prop('theme.colors.greyscale.dark')};

          &:hover,
          &:focus {
            color: ${({ theme }) => darken(0.1, theme.colors.greyscale.dark)};
          }
        `,
      )};`,
    hyperlink: css`
      text-decoration: underline;
      text-decoration-skip: ink;
      font-weight: ${prop('theme.fontWeight.regular')};
      color: ${prop('theme.colors.greyscale.black')};`,
    inherit: css`
      color: inherit;`,
  })};
`;

Anchor.defaultProps = {
  element: 'a',
  target: '_blank',
  variant: 'standard',
};

export default Anchor;
