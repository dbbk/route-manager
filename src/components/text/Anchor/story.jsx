// @flow

import React from 'react';
import { storiesOf } from '@storybook/react';

import Anchor from './';

const El = ({ to, children, className }) => (
  <abbr title={to} className={className}>
    {children}
  </abbr>
);

storiesOf('Components / Text / Anchor', module)
  .add('Standard - normal href', () => (
    <Anchor href="https://www.google.com">Google</Anchor>
  ))
  .add('Hyperlink - normal href', () => (
    <Anchor variant="hyperlink" href="https://www.google.com">
      Google
    </Anchor>
  ))
  .add('Navigation - normal href', () => (
    <Anchor variant="navigation" href="https://www.google.com">
      Google
    </Anchor>
  ))
  .add('with email prop', () => (
    <Anchor email="helpdesk@festicket.com">Helpdesk</Anchor>
  ))
  .add('with custom element', () => (
    <Anchor to="https://www.google.com" element={El} variant="standard">
      Google
    </Anchor>
  ));
