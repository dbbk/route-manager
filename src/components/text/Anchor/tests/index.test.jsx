// @flow

import * as React from 'react';
import { spy } from 'sinon';
import { mount, shallow } from 'enzyme';
import Anchor from '../';

// Simulate a custom link component, for example React Router's Link
function El({ children, to }: { children?: React.Node, to: string }) {
  return <span title={to}>{children}</span>;
}

describe('<Anchor />', () => {
  test('should render correctly', () => {
    const wrapper = mount(<Anchor href="http://www.google.com">Google</Anchor>);
    expect(wrapper).toMatchSnapshot();
  });

  test('should render correctly with custom element', () => {
    const wrapper = mount(
      <Anchor to="http://www.google.com" element={El}>
        Google
      </Anchor>,
    );
    expect(wrapper).toMatchSnapshot();
  });

  test('should render correctly with email prop', () => {
    const wrapper = mount(<Anchor email="email@test.com">Email me</Anchor>);
    expect(wrapper).toMatchSnapshot();
  });

  test('should render correctly with tel prop', () => {
    const wrapper = mount(<Anchor tel="08000337132">Phone us</Anchor>);
    expect(wrapper).toMatchSnapshot();
  });

  it('should call function if onClick property is passed', () => {
    const handle = spy();
    const component = shallow(
      <Anchor href="http://www.google.com" onClick={handle}>
        Google
      </Anchor>,
    );
    expect(handle.callCount).toEqual(0);
    component.simulate('click');
    expect(handle.callCount).toEqual(1);
  });

  it('should call function if onClick property is passed with email', () => {
    const handle = spy();
    const component = shallow(
      <Anchor email="email@test.com" onClick={handle}>
        Email me
      </Anchor>,
    );
    expect(handle.callCount).toEqual(0);
    component.simulate('click');
    expect(handle.callCount).toEqual(1);
  });

  it('should call function if onClick property is passed with external', () => {
    const handle = spy();
    const component = shallow(
      <Anchor href="http://www.google.com" onClick={handle} external>
        Google
      </Anchor>,
    );
    expect(handle.callCount).toEqual(0);
    component.simulate('click');
    expect(handle.callCount).toEqual(1);
  });
});
