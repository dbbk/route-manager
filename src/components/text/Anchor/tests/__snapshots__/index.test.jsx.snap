// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Anchor /> should render correctly 1`] = `
.c0 {
  cursor: pointer;
  text-decoration: none;
}

<Anchor
  href="http://www.google.com"
>
  <Styled(AnchorComponent)
    element="a"
    href="http://www.google.com"
    target="_blank"
    variant="standard"
  >
    <AnchorComponent
      className="c0"
      element="a"
      href="http://www.google.com"
      target="_blank"
      variant="standard"
    >
      <a
        className="c0"
        element="a"
        href="http://www.google.com"
        target="_blank"
        variant="standard"
      >
        Google
      </a>
    </AnchorComponent>
  </Styled(AnchorComponent)>
</Anchor>
`;

exports[`<Anchor /> should render correctly with custom element 1`] = `
.c0 {
  cursor: pointer;
  text-decoration: none;
}

<Anchor
  element={[Function]}
  to="http://www.google.com"
>
  <Styled(AnchorComponent)
    element={[Function]}
    target="_blank"
    to="http://www.google.com"
    variant="standard"
  >
    <AnchorComponent
      className="c0"
      element={[Function]}
      target="_blank"
      to="http://www.google.com"
      variant="standard"
    >
      <El
        className="c0"
        element={[Function]}
        target="_blank"
        to="http://www.google.com"
        variant="standard"
      >
        <span
          title="http://www.google.com"
        >
          Google
        </span>
      </El>
    </AnchorComponent>
  </Styled(AnchorComponent)>
</Anchor>
`;

exports[`<Anchor /> should render correctly with email prop 1`] = `
.c0 {
  cursor: pointer;
  text-decoration: none;
}

<Anchor
  email="email@test.com"
>
  <Styled(AnchorComponent)
    element="a"
    email="email@test.com"
    target="_blank"
    variant="standard"
  >
    <AnchorComponent
      className="c0"
      element="a"
      email="email@test.com"
      target="_blank"
      variant="standard"
    >
      <a
        className="c0"
        href="mailto:email@test.com"
      >
        Email me
      </a>
    </AnchorComponent>
  </Styled(AnchorComponent)>
</Anchor>
`;

exports[`<Anchor /> should render correctly with tel prop 1`] = `
.c0 {
  cursor: pointer;
  text-decoration: none;
}

<Anchor
  tel="08000337132"
>
  <Styled(AnchorComponent)
    element="a"
    target="_blank"
    tel="08000337132"
    variant="standard"
  >
    <AnchorComponent
      className="c0"
      element="a"
      target="_blank"
      tel="08000337132"
      variant="standard"
    >
      <a
        className="c0"
        href="tel:08000337132"
      >
        Phone us
      </a>
    </AnchorComponent>
  </Styled(AnchorComponent)>
</Anchor>
`;
